Makefile.docker

```diff
@@ -0,0 +1,128 @@
+	$(DOCKER_RUN) /bin/sh -c "make clean && $(CHOWN)"
+
+.PHONY:clean-images
+clean-images: ## Cleans the intermediate and final agent images left over from the build-image target
+	@# Clean any intermediate agent images, left over from the multi-stage build
+	$(call CLEAN_DOCKER_IMAGES,--filter label=com.logdna.stage=$(REPO)-build $(CLEAN_ADDTL_FILTERS))
+	@# Clean any final agent images
+	$(call CLEAN_DOCKER_IMAGES,--filter label=com.logdna.stage=$(REPO)-final $(CLEAN_ADDTL_FILTERS))
+
+.PHONY:build-image
+build-image: ## Build a docker image as specified in the Dockerfile
+	$(DOCKER) build . -t $(REPO):$(BUILD_TAG) \
+		$(PULL_OPTS) --no-cache=true --rm \
+		--build-arg BUILD_IMAGE=$(IMAGE) \
+		--build-arg BUILD_TIMESTAMP=$(BUILD_TIMESTAMP) \
+		--build-arg BUILD_VERSION=$(BUILD_VERSION) \
+		--build-arg REPO=$(REPO) \
+		--build-arg VCS_REF=$(VCS_REF) \
+		--build-arg VCS_URL=$(VCS_URL)
+	$(DOCKER) tag $(REPO):$(BUILD_TAG) $(REPO):$(BUILD_VERSION)
+
+.PHONY:publish-private
+publish-private: ## Publish a build version of the docker image to our private registry
+	for version in $(BUILD_TAG) $(BUILD_VERSION); do \
+		$(DOCKER) tag $(REPO):$(BUILD_TAG) $(DOCKER_PRIVATE_IMAGE):$${version}; \
+		$(DOCKER) push $(DOCKER_PRIVATE_IMAGE):$${version}; \
+	done;
+
+.PHONY:publish-public
+publish-public: ## Publish SemVer compliant releases to our public registries
+	@# TODO: Have a boolean that prevents this unless forced or run by Jenkins (which can force it)
+	for image in $(DOCKER_PUBLIC_IMAGE) $(DOCKER_IBM_IMAGE); do \
+		for version in $(MAJOR_VERSION) $(MINOR_VERSION) $(PATCH_VERSION) $(BUILD_VERSION) latest; do \
+			if [[ $${version} != $(BUILD_VERSION) ]] && [[ $(PRERELEASE_VERSION) = beta* ]]; then \
+				continue; \
+			fi; \
+			$(DOCKER) tag $(REPO):$(BUILD_TAG) $${image}:$${version}; \
+			$(DOCKER) push $${image}:$${version}; \
+		done; \
+	done;
+
+.PHONY:get-rust-image
+get-rust-image:
+	@echo $(RUST_REPO):$(BUILD_TAG)
+
+.PHONY:rust-image
+rust-image: ## Build a superset of the Rust image that contains necessary dependencies to test the agent
+	$(DOCKER) build . -t $(RUST_REPO):$(BUILD_TAG) \
+		-f rust-image/Dockerfile \
+		--pull --no-cache=true --rm \
+		--build-arg BUILD_IMAGE=$(IMAGE) \
+		--build-arg REPO=$(RUST_REPO) \
+		--build-arg VCS_REF=$(VCS_REF)
+
+.PHONY:publish-rust
+publish-rust: ## Publish superset rust image (see target rust-image)
+	$(DOCKER) tag $(RUST_REPO):$(BUILD_TAG) docker.pkg.github.com/answerbook/docker-images/$(RUST_REPO):latest
+	$(DOCKER) tag $(RUST_REPO):$(BUILD_TAG) us.gcr.io/logdna-k8s/$(RUST_REPO):latest
+	$(DOCKER) push docker.pkg.github.com/answerbook/docker-images/$(RUST_REPO):latest
+	$(DOCKER) push us.gcr.io/logdna-k8s/$(RUST_REPO):latest
+
+.PHONY:clean-rust-image
+clean-rust-image:
+	$(call CLEAN_DOCKER_IMAGES,--filter label=com.logdna.stage=$(RUST_REPO)-build $(CLEAN_ADDTL_FILTERS))
+	$(call CLEAN_DOCKER_IMAGES,--filter label=com.logdna.stage=$(RUST_REPO)-build $(CLEAN_ADDTL_FILTERS))
```

Comment:
If we are going to maintain a base rust image, this should really be external to this repository in its own.

Reply:
I was on the fence about how to handle this. I was originally going to add the logic for building the logdna-agent rust base image to https://github.com/answerbook/docker-images.

I decided to go with the current approach for the following reasons:
1. The base image is tightly coupled with the agent itself; breaking them apart doesn't remove the coupling but instead hides it
2. Kills some of the automagic. Changes on build or test dependencies would need the base image to be updated which would need to be manually done since it wouldn't be able to detect changes from a different repository.

The current approach does the following when a Jenkins build goes for the logdna-agent-v2:
1. Check for changes to the base image
1.a If there are changes build a new base image and use the local version for future steps
1.b Otherwise, use the remote image for future steps
2. Test / build the agent with whatever base image assigned previously
3. If successful and there's a new base image, push that image to the remote repo

It does admittedly make the Jenkins build script a bit more complicated. Thoughts?

Reply:
So my first question is... why are we building this rust base image ?

If the answer is that we absolutely have to, then i do think it should be in a separate repository. i would do a repository specifically for rust, because the problem you are describing is explicitly an issue we have to resolve regardless.  Over in `docker-images` we'll need to be updating these images regularly for security reasons, and triggering builds on things that haven't gotten a recent release (really it will just be blanket) so that we have updates available.

Increasing the complexity in this repository either way, to me, seems like the wrong route.

Reply:
I removed this and moved it to our private repo docker-images